app/src/org/commcare/dalvik/utils/UriToFilePath.java

```diff
@@ -0,0 +1,143 @@
+package org.commcare.dalvik.utils;
+
+import android.annotation.SuppressLint;
+import android.content.ContentUris;
+import android.content.Context;
+import android.database.Cursor;
+import android.net.Uri;
+import android.os.Build;
+import android.os.Environment;
+import android.provider.DocumentsContract;
+import android.provider.MediaStore;
+
+/**
+ * Util for turning URIs into a normal file path strings Needed because new
+ * versions of Android have a content resolution abstraction layer that doesn't
+ * use vanilla file system paths.
+ *
+ * Taken from aFileChooser by Paul Burke (paulburke.co)
+ * https://github.com/iPaulPro/aFileChooser
+ * Which is under Apache 2.0 license
```

Comment:
Since we're also on Apache2.0 there's no need to include any licenses in
the file directly.

There are two options:

1) Inline like this is totally fine and also helps people understand code
provenance so I generally like this.  (especially for full classes)

2) You can note in the root of the repository that code was taken from
specific sources.

-Clayton

On Fri, Apr 17, 2015 at 4:34 PM, Phillip Mates <notifications@github.com>
wrote:

> In app/src/org/commcare/dalvik/utils/UriToFilePath.java
> <https://github.com/dimagi/commcare-odk/pull/232#discussion_r28627912>:
>
> > +import android.content.Context;
> > +import android.database.Cursor;
> > +import android.net.Uri;
> > +import android.os.Build;
> > +import android.os.Environment;
> > +import android.provider.DocumentsContract;
> > +import android.provider.MediaStore;
> > +
> > +/**
> > + * Util for turning URIs into a normal file path strings Needed because new
> > + * versions of Android have a content resolution abstraction layer that doesn't
> > + * use vanilla file system paths.
> > + *
> > + * Taken from aFileChooser by Paul Burke (paulburke.co)
> > + * https://github.com/iPaulPro/aFileChooser
> > + * Which is under Apache 2.0 license
>
> How do we handle snippets of code taken from repos with Apache 2.0
> licenses?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-odk/pull/232/files#r28627912>.
>